src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,259 @@
+<?php
+
+class WP_HTML_Processor extends WP_HTML_Tag_Processor {
+	public $fully_supported_input = null;
+	public $open_elements = array();
+
+	public function ensure_support() {
+		if ( null !== $this->fully_supported_input ) {
+			return $this->fully_supported_input;
+		}
+
+		$stack = array();
+
+		$p = new WP_HTML_Tag_Processor( $this->html );
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			$tag_name = $p->get_tag();
+
+			if ( ! $p->is_tag_closer() ) {
+				$element = WP_HTML_Spec::element_info( $tag_name );
+
+				$self_closes = $element::is_void || ( ! $element::is_html && $p->has_self_closing_flag() );
+				if ( ! $self_closes ) {
+					$stack[] = $tag_name;
+				}
+			} else {
+				if ( end( $stack ) === $tag_name ) {
+					array_pop( $stack );
+					continue;
+				}
+
+				$this->fully_supported_input = false;
+				return false;
+			}
+		}
+
+		$this->fully_supported_input = 0 === count( $stack );
+
+		return $this->fully_supported_input;
+	}
+
+	public function next_tag( $query = null ) {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		if ( 0 < count( $this->open_elements ) ) {
+			$element = WP_HTML_Spec::element_info( end( $this->open_elements ) );
+			// @TODO: Handle self-closing HTML foreign elements: must convey self-closing flag on stack.
+			if ( $element::is_void ) {
+				array_pop( $this->open_elements );
```

Comment:
Shouldn't it say `if ( ! element::is_void )`?

Reply:
Or not? I don't get why void elements have any interaction with the open elements stack

Reply:
Oh I see, why are void elements landing on the open elements stack, though?

Reply:
this is something to document later and it appeared towards the end of the day. I think I've seen this in the HTML spec for tree-building.

the problem is that we have to push the void element to the open stack in order to get proper depth information and climb out of the nesting we're in. this looks funny here because we aren't currently tracking if the last element on the stack requires no closer (is void or self-closing foreign element). that means whenever we approach the next tag we have to immediately remove any void elements because they cannot enclose over other tags.

it's definitely unexpected until you reason through it, or as I did, get reminded by it with failing tests.

Reply:
According to the spec, popping should happen in the same step as the insertion:

> A start tag whose tag name is one of: "area", "br", "embed", "img", "keygen", "wbr"
[Reconstruct the active formatting elements](https://html.spec.whatwg.org/multipage/parsing.html#reconstruct-the-active-formatting-elements), if any.
> 
> [Insert an HTML element](https://html.spec.whatwg.org/multipage/parsing.html#insert-an-html-element) for the token. **Immediately pop the [current node](https://html.spec.whatwg.org/multipage/parsing.html#current-node) off the [stack of open elements](https://html.spec.whatwg.org/multipage/parsing.html#stack-of-open-elements).**
> 
> [Acknowledge the token's self-closing flag](https://html.spec.whatwg.org/multipage/parsing.html#acknowledge-self-closing-flag), if it is set.
>
> Set the [frameset-ok flag](https://html.spec.whatwg.org/multipage/parsing.html#frameset-ok-flag) to "not ok".